### ISSUE TITLE
future_step?("wicked_finished") false

### ISSUE THREAD

**Issue by nruth** [schneems/wicked#210]
This can cause problems in before-filters.

I'm not sure exactly what the "protected steps" do at the moment, but it seems like `step_index_for` should do something like this:

```rb
def index_of_step(step)
  if step == "wicked_finished".freeze
    # 0-indexed so no +1 needed
    wizard_steps.length
  else
    wizard_steps.index(step)
  end
end
```

Or would this break things unexpectedly for people?

**Comment by nruth** [comment]
This also affects e.g. `past_step?(:pay)` which should be true for wicked_finished but is currently false.

**Comment by schneems** [comment]
I think this is because we set those values in a before filter. We should be appending and setting it before any other filters but it might be that it's not happening in that order in this case.

**Comment by nruth** [comment]
Thanks for the feedback. I'll try to make time to look at this soon, but if the issue goes stale please feel free to close it.

The things I'd do are:
- Write some tests demonstrating what I expected
- Look more at how things work, and figure out whether filters or the next/previous step query methods should change

**Comment by schneems** [comment]
I guess i'm also curious as to why you're using `wicked_finished` inside of future_step? having more context around that use case would help me.

**Comment by nruth** [comment]
Yes that does look a bit crazy doesn't it.


I don't think I was actually writing `future_step?("wicked_finished")`, I just used it as the ticket title to show quickly what it was about. I was actually writing something like `past_step?(:pay)` against these steps:

```
steps :register_email, :confirm_email, :receive_pin, :pay, :payment_received_set_exam_date
```

Then I had some problem with a filter firing on the final step (`wicked_finished`) when I expected it not to fire. I don't remember if this prevented the form saving data or just gave an unexpected redirect.

`future_step?("wicked_finished")` was one of the things I tried out while playing around in pry / rails console, and in haste I used it as the issue title by mistake.

I'm finding it hard to write a good description for this without getting back into the code again and reproducing the problem in my controller (I was trying things out and either committed a solution checking the steps another way, or left that rare edge-case for later.


I'm sorry this is pretty vague at the moment, but I hope it's a bit clearer what I was doing when I saw this.
